com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl

```diff
@@ -169,164 +169,296 @@ float4 SampleEnv(LightLoopContext lightLoopContext, int index, float3 texCoord,
 // Single Pass and Tile Pass
 // ----------------------------------------------------------------------------
 
-#ifndef LIGHTLOOP_DISABLE_TILE_AND_CLUSTER
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/TilingAndBinningUtilities.hlsl"
 
-// Calculate the offset in global light index light for current light category
-int GetTileOffset(PositionInputs posInput, uint lightCategory)
-{
-    uint2 tileIndex = posInput.tileCoord;
-    return (tileIndex.y + lightCategory * _NumTileFtplY) * _NumTileFtplX + tileIndex.x;
-}
+#if (defined(COARSE_BINNING) || defined(FINE_BINNING))
 
-void GetCountAndStartTile(PositionInputs posInput, uint lightCategory, out uint start, out uint lightCount)
+// Internal. Do not call directly.
+uint TryFindEntityIndex(inout uint i, uint tile, uint2 zBinRange, uint category, out uint entityIndex)
 {
-    int tileOffset = GetTileOffset(posInput, lightCategory);
+    entityIndex = UINT16_MAX;
 
-#if defined(UNITY_STEREO_INSTANCING_ENABLED)
-    // Eye base offset must match code in lightlistbuild.compute
-    tileOffset += unity_StereoEyeIndex * _NumTileFtplX * _NumTileFtplY * LIGHTCATEGORY_COUNT;
-#endif
+    bool b = false;
+    uint n = TILE_ENTRY_LIMIT;
 
-    // The first entry inside a tile is the number of light for lightCategory (thus the +0)
-    lightCount = g_vLightListGlobal[DWORD_PER_TILE * tileOffset + 0] & 0xffff;
-    start = tileOffset;
-}
+    // This part (1) is loop-invariant. These values do not depend on the value of 'i'.
+    // They will only be computed once per category, not once per function call.
+    const uint tileBufferHeaderIndex = ComputeTileBufferHeaderIndex(tile, category, unity_StereoEyeIndex);
+    const uint tileRangeData         = TILE_BUFFER[tileBufferHeaderIndex]; // {last << 16 | first}
+    const bool isTileEmpty           = tileRangeData == UINT16_MAX;
 
-#ifdef USE_FPTL_LIGHTLIST
+    if (!isTileEmpty) // Avoid wasted work
+    {
+        const uint zBinBufferIndex0 = ComputeZBinBufferIndex(zBinRange[0], category, unity_StereoEyeIndex);
+        const uint zBinBufferIndex1 = ComputeZBinBufferIndex(zBinRange[1], category, unity_StereoEyeIndex);
+        const uint zBinRangeData0   = _zBinBuffer[zBinBufferIndex0]; // {last << 16 | first}
+        const uint zBinRangeData1   = _zBinBuffer[zBinBufferIndex1]; // {last << 16 | first}
+
+        // Recall that entities are sorted by the z-coordinate.
+        // So we can take the smallest index from the first bin and the largest index from the last bin.
+        // TODO: while it appears to work, for a large range, this code can actually fail. :-(
```

Comment:
Question: Why don't you do two functions? one for the case of a single position (i.e we use in light loop) 
and thus const uint2 zBinEntityIndexRange = uint2(zBinRangeData0 & UINT16_MAX, zBinRangeData1 >> 16); = >onst uint2 zBinEntityIndexRange = uint2(zBinRangeData & UINT16_MAX, zBinRangeData >> 16);

and another one for the volumetric with the range. Are you trusting the compiler to do the right thing here?

Reply:
This optimization is trivial. I don't want to duplicate the code.